File: Java/src/com/company/Array/LC238_productExceptSelf.java
package com.company.Array;

public class LC238_productExceptSelf {
    public int[] productExceptSelf(int[] n) {
        /*
            explaination

            n :    1      2   3       4
        left  :    1      1  1*2*3  1*2*3*4
        right : 4*3*2*1  4*3  4*1      1
        */
        int[] left = new int[n.length];
        left[0] = 1;
        for(int i = 1; i < left.length; i++){
            left[i] = left[i-1] * n[i-1];
        }

        int[] res = new int[n.length];
        int right = 1;
        for(int i = n.length-1; i >= 0; i--){
            res[i] = left[i] * right;
            right = right * n[i];
        }
        return res;
    }
}

File: python/Array/LC238_productExceptSelf.py
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # imagine when looping from left to right, in a certain view, we can only know the product of the values on the left
        # in order to knnow the product of values on the right from this view, just loop from right to left again
        
        # space O(1)
#         rtn = [1] * len(nums)
#         prod = 1        
#         for i in range(len(nums)):
#             rtn[i] *= prod
#             prod *= nums[i]
        
#         prod = 1
#         for i in range(len(nums)-1, -1, -1):
#             rtn[i] *= prod
#             prod *= nums[i]
            
#         return rtn
        
        # space O(n)
        left = [0] * len(nums)
        right = [0] * len(nums)
        left[0] = 1
        right[-1] = 1
        for i in range(1, len(nums)):
            left[i] = left[i-1] * nums[i-1]
            
        for i in range(len(nums)-2, -1, -1):
            right[i] = right[i+1] * nums[i+1]
        
        
        rtn = [left[i] * right[i] for i in range(len(nums))]
        return rtn
